### PR TITLE
Migrate codebase to support py36 and above

### PR DESCRIPTION
**Pull request by ljvmiranda921**
## Description

This PR attempts to update the wasabi codebase so that it's compatible to Python 3.6 and above.
Major changes include:
- Added type hints to all modules
- Removed Python27 from the CI
- Ran `isort -m 9` and `black` to format the codebase
- Added `mypy` to `requirements.txt`, and included it in the CI.

### A note on typing and mypy

There are areas in the codebase where I prioritized readability than `mypy`'s checks, By doing so, there are some lines of code where I added `type: ignore` or `cast`ed some variables to the appropriate types. Let me know if there's a better way to approach these. Here are some examples:

- Ignored type when using `**kwargs`: https://github.com/ines/wasabi/pull/22/commits/f8f3c595afad55e2fa39366f88541ae88d21f8d1. Here, I prioritized neatness (I prefer the original `**settings` approach) rather than passing the variables one by one.
- Fix dictionary getters by casting: https://github.com/ines/wasabi/pull/22/commits/534311653f2f6b0b73b1a6d363cb5f9190d0d8ed. Here, I opted to cast the variables instead of writing overloads.

Also, note that I'm not *as* familiar with `typing` and `mypy` best practices, so let me know if there are things I missed! Would appreciate some pointers!

### CODE REVIEW

**Comment by ljvmiranda921** [comment]
Yay ready for review! Can't seem to tag reviewers so sorry for the message tag cc: @ines

**Comment by svlandeg** [comment]
Does the CI run `mypy` in the test suite?

File: azure-pipelines.yml [hunk 1]
```diff
@@ -79,14 +61,16 @@ jobs:
...
+      python -m pip install --upgrade pip
+      python -m pip install -r requirements.txt
     displayName: 'Install dependencies'
 
   - script: |
       python -c "import locale; print('Preferred encoding:', locale.getpreferredencoding())"
       python -c "import sys; print('stdout encoding:', sys.stdout.encoding)"
       env
-      pip install pytest
+      python -m mypy wasabi
```
Comment (adrianeboyd) [inline azure-pipelines.yml:73]: Split this into two separate steps so a mypy failure leads to a clear CI failure. (I'm pretty sure it uses the exit code of the final command in the step.)

**Comment by adrianeboyd** [comment]
Please fix CI steps first and address the mypy errors (which are currently printed in the CI but don't cause the step to fail).

**Comment by ljvmiranda921** [comment]
Oh my mistake @adrianeboyd , I re-requested the review but  forgot to push the changes! I have pushed them now and will wait for the CI to finish. Once everything passes, I will request the review again. :) I apologize for that.

File: azure-pipelines.yml (hunk 1 above)
Comment (ljvmiranda921) [reply azure-pipelines.yml]: Updated in https://github.com/explosion/wasabi/pull/22/commits/590c00847b239bf59a142571f8bc1f7354097b20

**Comment by adrianeboyd** [comment]
You need to merge in `master` and resolve the conflicts for the CI to run again.

**Comment by adrianeboyd** [comment]
Ah, the CI hasn't been updated since it got moved to explosion so it's not working. Let me try to fix it...

**Comment by ljvmiranda921** [comment]
Ok! I reran `mypy` locally and encountered two errors:

```
λ: python3 -m mypy wasabi
wasabi/util.py:210: error: Module has no attribute "WinDLL"
wasabi/util.py:230: error: Module has no attribute "get_osfhandle"
Found 2 errors in 1 file (checked 13 source files)
```

I will solve these first in the meantime 🙇

**Comment by adrianeboyd** [comment]
That code should have been removed in the merge? It doesn't matter to me if this PR gets rebased, so it might be easier to revert back to before the merge and rebase on master instead?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -79,16 +61,24 @@ jobs:
...
       python -c "import sys; print('stdout encoding:', sys.stdout.encoding)"
       env
-      pip install pytest
       python -m pytest -s wasabi
     displayName: 'Run tests'
 
+  - script: |
+      python -c "import locale; print('Preferred encoding:', locale.getpreferredencoding())"
+      python -c "import sys; print('stdout encoding:', sys.stdout.encoding)"
+      env
```
Comment (adrianeboyd) [inline azure-pipelines.yml:79]: This is just debugging and it buries the actual interesting part of the logs.

```suggestion
```

**Comment by ljvmiranda921** [comment]
I noticed that closing and opening the PR re-triggers the CI. Earlier when I made a commit, the CI didn't trigger.

File: wasabi/markdown.py [hunk 3]
```diff
@@ -1,36 +1,44 @@
...
+    def table(
+        self,
+        data: List[List[str]],
+        header: List[str],
+        aligns: List[Literal["r", "c", "l"]] = None,
+    ) -> str:
         """Create a Markdown table.
 
-        data (Iterable[Iterable[str]]): The body, one iterable per row,
+        data (List[List[str]]): The body, one iterable per row,
```
Comment (adrianeboyd) [inline wasabi/markdown.py:37]: Why is this changed?

File: wasabi/printer.py [hunk 4]
```diff
@@ -187,10 +215,10 @@ def divider(self, text="", char="=", show=True, icon=None):
             return text
         print(text)
 
-    def table(self, data, **kwargs):
+    def table(self, data: Union[List, Dict], **kwargs):
         """Print data as a table.
 
-        data (iterable / dict): The data to render. Either a list of lists
+        data (Iterable / Dict): The data to render. Either a list of lists
```
Comment (adrianeboyd) [inline wasabi/printer.py:221]: This looks like it's out-of-sync with the defined types?

File: wasabi/printer.py [hunk 5]
```diff
@@ -202,10 +230,10 @@ def table(self, data, **kwargs):
             return text
         print(text)
 
-    def row(self, data, **kwargs):
+    def row(self, data: List, **kwargs):
         """Print a table row.
 
-        data (iterable): The individual columns to format.
+        data (Iterable): The individual columns to format.
```
Comment (adrianeboyd) [inline wasabi/printer.py:236]: Also out-of-sync?

**Review by adrianeboyd — COMMENTED** [review]
Can you double-check all the definitions vs. docstrings again?

I think that it makes sense to continue supporting the data as `Iterable` in a lot of cases, but maybe there's some aspect I've missed?

**Comment by ljvmiranda921** [comment]
Hi @adrianeboyd , I think for the change from `Iterable` to `List`, I had a few `mypy` errors that say something like:

```
wasabi/markdown.py:45: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
wasabi/markdown.py:46: error: Argument 1 to "len" has incompatible type "Iterable[Literal['r', 'c', 'l']]"; expected "Sized"
wasabi/markdown.py:46: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
wasabi/markdown.py:47: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
wasabi/markdown.py:52: error: Value of type "Iterable[Literal['r', 'c', 'l']]" is not indexable
wasabi/markdown.py:52: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
```

I am not sure what the best type for that is. I searched a bit (cf. [this SO post](https://stackoverflow.com/questions/49427944/typehints-for-sized-iterable-in-python)), and it mentions that a `List`, `Collection` or `Sequence` should work. I opted to use `List` because we're passing a `list` internally.

**Comment by ljvmiranda921** [comment]
Noted on the docstrings, will double check them again.

**Comment by ljvmiranda921** [comment]
Converting this to draft again to just double-check all docstrings.

File: wasabi/markdown.py (hunk 3 above)
Comment (ljvmiranda921) [reply wasabi/markdown.py]: Hi @adrianeboyd , I think for the change from Iterable to List, I had a few mypy errors that say something like:

```
wasabi/markdown.py:45: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
wasabi/markdown.py:46: error: Argument 1 to "len" has incompatible type "Iterable[Literal['r', 'c', 'l']]"; expected "Sized"
wasabi/markdown.py:46: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
wasabi/markdown.py:47: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
wasabi/markdown.py:52: error: Value of type "Iterable[Literal['r', 'c', 'l']]" is not indexable
wasabi/markdown.py:52: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
```

I am not sure what the best type for that is. I searched a bit (cf. [this SO post](https://stackoverflow.com/questions/49427944/typehints-for-sized-iterable-in-python)), and it mentions that a List, Collection or Sequence should work. I opted to use List because we're passing a list internally.

File: wasabi/__init__.py [hunk 6]
```diff
@@ -1,12 +1,12 @@
 # coding: utf8
 from __future__ import unicode_literals
 
+from .about import __version__  # noqa
+from .markdown import MarkdownRenderer  # noqa
 from .printer import Printer  # noqa
-from .tables import table, row  # noqa
+from .tables import row, table  # noqa
```
Comment (ljvmiranda921) [inline wasabi/__init__.py:7]: Some of these changes were done via `isort`

File: wasabi/markdown.py [hunk 7]
```diff
@@ -1,36 +1,44 @@
...
+        data: Iterable[Iterable[str]],
+        header: Sequence[str],
+        aligns: Optional[Sequence[Literal["r", "c", "l"]]] = None,
+    ) -> str:
         """Create a Markdown table.
 
         data (Iterable[Iterable[str]]): The body, one iterable per row,
             containig an interable of column contents.
-        header (Iterable[str]): The column names.
+        header (Sequence[str]): The column names.
```
Comment (ljvmiranda921) [inline wasabi/markdown.py:39]: Here, the `Iterable` type causes a `mypy` error:

```
wasabi/markdown.py:45: error: Argument 1 to "len" has incompatible type "Iterable[str]"; expected "Sized"
```

So I changed it into a `Sequence`.

File: wasabi/printer.py [hunk 8]
```diff
@@ -67,67 +65,127 @@ def __init__(
...
 
-        title (unicode): The main text to print.
-        text (unicode): Optional additional text to print.
-        color (unicode / int): Foreground color.
-        bg_color (unicode / int): Background color.
-        icon (unicode): Name of icon to add.
-        spaced (unicode): Whether to add newlines around the output.
+        title (str): The main text to print.
+        text (str): Optional additional text to print.
+        color (str / int): Optional foreground color.
```
Comment (ljvmiranda921) [inline wasabi/printer.py:174]: I think it's much more readable to write `(str / int)` rather than `Union[str, int]` in the docstring.

File: wasabi/printer.py [hunk 9]
```diff
@@ -67,67 +65,127 @@ def __init__(
...
         """
         if not show:
             return
         if self.pretty:
-            color = self.colors.get(color, color)
-            bg_color = self.colors.get(bg_color, bg_color)
-            icon = self.icons.get(icon)
+            color = self.colors.get(cast(str, color), color)
+            bg_color = self.colors.get(cast(str, bg_color), bg_color)
+            icon = self.icons.get(cast(str, icon))
```
Comment (ljvmiranda921) [inline wasabi/printer.py:188]: There are some value re-assignment happening here. So I opted to just type-cast the variable rather than creating a new one.

File: wasabi/tables.py [hunk 10]
```diff
@@ -78,45 +79,45 @@ def table(
         "fg_colors": fg_colors,
         "bg_colors": bg_colors,
     }
-    divider_row = row(["-" * width for width in widths], **settings)
+    divider_row = row(["-" * width for width in widths], **settings)  # type: ignore
```
Comment (ljvmiranda921) [inline wasabi/tables.py:82]: I'm not familiar how to do good typing when using `**kwargs`. So for this I opted to just ignore the type.

File: wasabi/tables.py [hunk 11]
```diff
@@ -78,45 +79,45 @@ def table(
...
         default), 'r' (right) or 'c' (center). If a string, value is used
         for all columns.
-    env_prefix (unicode): Prefix for environment variables, e.g.
+    env_prefix (str): Prefix for environment variables, e.g.
         WASABI_LOG_FRIENDLY.
-    fg_colors (list): Foreground colors for the columns, in order. None can be
+    fg_colors (list): Optional foreground colors for the columns, in order. None can be
         specified for individual columns to retain the default foreground color.
-    bg_colors (list): Background colors for the columns, in order. None can be
+    bg_colors (list): Optional background colors for the columns, in order. None can be
```
Comment (ljvmiranda921) [inline wasabi/tables.py:120]: When a type is `Optional`, I wrote the docstring where the first word starts with "Optional...".

File: wasabi/tables.py [hunk 12]
```diff
@@ -126,16 +127,17 @@ def row(
         and not env_log_friendly
         and (fg_colors is not None or bg_colors is not None)
     )
-    cols = []
-    if isinstance(aligns, basestring_):  # single align value
-        aligns = [aligns for _ in data]
+    cols: List[str] = []
+    _aligns = (
+        [aligns for _ in data] if isinstance(aligns, str) else cast(List[str], aligns)
```
Comment (ljvmiranda921) [inline wasabi/tables.py:132]: There are some value re-assignment here, so I opted to just type-cast the `aligns` method and reassign its value to another variable.

File: wasabi/printer.py [hunk 13]
```diff
@@ -187,10 +247,10 @@ def divider(self, text="", char="=", show=True, icon=None):
             return text
         print(text)
 
-    def table(self, data, **kwargs):
+    def table(self, data: Union[Collection, Dict], **kwargs):
         """Print data as a table.
 
-        data (iterable / dict): The data to render. Either a list of lists
+        data (Collection / Dict): The data to render. Either a list of lists
```
Comment (ljvmiranda921) [inline wasabi/printer.py:253]: The `Iterable` type causes an error here:

```
# The `data` variable is passed on to wasabi.table
wasabi/tables.py:69: error: Argument 1 to "len" has incompatible type "Iterable[Any]"; expected "Sized"
```

I decided to use `Collection` since I don't think there's no need for this data to be ordered.

**Review by ljvmiranda921 — COMMENTED** [review]
Hi @adrianeboyd , thank you for the review! I checked the typing and updated a few docstrings as well.
I also tried to annotate some of my changes within the lines of code (hope Github shows them properly).

File: wasabi/markdown.py (hunk 7 above)
Comment (adrianeboyd) [reply wasabi/markdown.py:39]: I think the `List` is okay for `header`, I was mainly concerned about `data` and whether we potentially had `Iterable` uses for large tables.

File: wasabi/__init__.py [hunk 14]
```diff
@@ -1,12 +1,12 @@
 # coding: utf8
 from __future__ import unicode_literals
```
Comment (adrianeboyd) [inline wasabi/__init__.py:3]: There are also still some remaining `coding` lines in a couple places?
```suggestion
```

Comment (ljvmiranda921) [reply wasabi/__init__.py:3]: Oh okay I missed that. Will update them in a few!

File: wasabi/tables.py (hunk 11 above)
Comment (adrianeboyd) [reply wasabi/tables.py:120]: Let's use the actual types in the docstrings rather than relying on the description?

File: wasabi/markdown.py (hunk 7 above)
Comment (adrianeboyd) [reply wasabi/markdown.py:39]: `Sequence` is fine for `header`, too.

File: wasabi/tables.py (hunk 11 above)
Comment (ljvmiranda921) [reply wasabi/tables.py:120]: Ok will do that! There's also a type of docstring where if the input type is `Union[TypeA, TypeB]`, then I write `param (type_a / type_b)` in the docstring. I guess for clarity, I will copy over what's in the docstring?

Comment (adrianeboyd) [reply wasabi/tables.py:120]: I think `(str / int)` is okay, but `Optional` has a specific meaning as a type that's not as clear as part of the description.

**Comment by ljvmiranda921** [comment]
Hi @adrianeboyd ! I'm not sure if I have pending changes for the PR. I tried re-opening some resolved threads, perhaps a comment was buried or deleted somewhere? Let me know if there are things I need to update or if I should click the "request re-review" button! 🙇

File: wasabi/tables.py (hunk 11 above)
Comment (ljvmiranda921) [reply wasabi/tables.py]: Oh I just saw your last comment! Oops, I've already changed them to `Union[A, B]` in the docstrings (https://github.com/explosion/wasabi/pull/22/commits/1ef7d56e08c06a2b111f46c57d5eaf0e9c42b221). I can revert them back! Let me know if I should do that instead.

**Comment by ljvmiranda921** [comment]
> Ah, with Optional[] I do like the Union version better. Maybe Union everywhere would just be simpler/clearer?

Ah sure. I think Union everywhere is clearer. Will revert this commit: https://github.com/explosion/wasabi/pull/22/commits/fb2c728aa5d095f7b0d4a89cceb55adaf3926fc7

> An additional change would be to add python_version=">=3.6" to the setup.

Sure I can add that

File: wasabi/tables.py [hunk 15]
```diff
@@ -1,54 +1,56 @@
...
-    bg_colors=None,
-):
+    data: Union[Collection, Dict],
+    header: Optional[Iterable] = None,
+    footer: Optional[Iterable] = None,
+    divider: bool = False,
+    widths: Union[Iterable[int], Literal["auto"]] = "auto",
+    max_col: int = 30,
+    spacing: int = 3,
+    aligns: Optional[Union[Iterable[Literal["r", "c", "l"]], str]] = None,
```
Comment (adrianeboyd) [inline wasabi/tables.py:23]: Technically this is 

```python
Optional[Union[Iterable[Literal["r", "c", "l"]], Literal["r", "c", "l"]]
```

?

File: wasabi/tables.py [hunk 16]
```diff
@@ -1,54 +1,56 @@
...
     divider (bool): Show a divider line between header/footer and body.
-    widths (iterable or 'auto'): Column widths in order. If "auto", widths
+    widths (Union[Iterable[int], Literal['auto']]): Column widths in order. If "auto", widths
         will be calculated automatically based on the largest value.
     max_col (int): Maximum column width.
     spacing (int): Spacing between columns, in spaces.
-    aligns (iterable / unicode): Column alignments in order. 'l' (left,
-        default), 'r' (right) or 'c' (center). If a string, value is used
-        for all columns.
+    aligns (Optional[Union[Iterable[Literal['r', 'c', 'l']], str]]): Optional column alignments
```
Comment (adrianeboyd) [inline wasabi/tables.py:41]: For the docstring, maybe just?

```suggestion
    aligns (Optional[Union[Iterable[str], str]]): Optional column alignments
```

**Comment by adrianeboyd** [comment]
Looks good!
